cpp/daal/src/algorithms/dtrees/gbt/regression/gbt_regression_predict_dense_default_batch_impl.i

```diff
@@ -220,26 +291,320 @@ services::Status PredictRegressionTask<algorithmFPType, cpu>::run(const gbt::reg
+    //   - the tree prediction for leaf nodes
+    // we are accounting for the fact that nodeIndex is counted from 1 (not from 0) as required by helper functions
+    const gbt::prediction::internal::ModelFPType * const splitValues     = tree->getSplitPoints() - 1;
+    const gbt::prediction::internal::FeatureIndexType * const fIndexes   = tree->getFeatureIndexesForSplit() - 1;
+    const gbt::prediction::internal::ModelFPType * const nodeCoverValues = tree->getNodeCoverValues() - 1;
+    const int * const defaultLeft                                        = tree->getDefaultLeftForSplit() - 1;
+
+    // stop if we have no weight coming down to us
+    if (conditionFraction == 0) return;
+
+    // extend the unique path
+    PathElement * uniquePath = parentUniquePath + uniqueDepth + 1;
+    const size_t nBytes      = uniqueDepth * sizeof(PathElement);
+    const int copy_status    = daal::services::internal::daal_memcpy_s(uniquePath, nBytes, parentUniquePath, nBytes);
+    DAAL_ASSERT(copy_status == 0);
+
+    // std::copy(parentUniquePath, parentUniquePath + uniqueDepth + 1, uniquePath);
+
+    if (condition == 0 || conditionFeature != parentFeatureIndex)
+    {
+        extendPath(uniquePath, uniqueDepth, parentZeroFraction, parentOneFraction, parentFeatureIndex);
+    }
+
+    const bool isLeaf = gbt::internal::ModelImpl::nodeIsLeaf(nodeIndex, *tree, depth);
+
+    // leaf node
+    if (isLeaf)
+    {
+        for (std::uint32_t i = 1; i <= uniqueDepth; ++i)
+        {
+            const float w          = unwoundPathSum(uniquePath, uniqueDepth, i);
+            const PathElement & el = uniquePath[i];
+            phi[el.featureIndex] += w * (el.oneFraction - el.zeroFraction) * splitValues[nodeIndex] * conditionFraction;
+        }
+
+        return;
+    }
+
+    const gbt::prediction::internal::FeatureIndexType splitFeature = fIndexes[nodeIndex];
+    const auto dataValue                                           = x[splitFeature];
+
+    dispatcher_t<hasUnorderedFeatures, hasAnyMissing> dispatcher;
+    gbt::prediction::internal::FeatureIndexType hotIndex =
+        updateIndex(nodeIndex, dataValue, splitValues, defaultLeft, _featHelper, splitFeature, dispatcher);
+    const gbt::prediction::internal::FeatureIndexType coldIndex = 2 * nodeIndex + (hotIndex == (2 * nodeIndex));
+
+    const float w = nodeCoverValues[nodeIndex];
+    DAAL_ASSERT(w > 0);
+    const float hotZeroFraction  = nodeCoverValues[hotIndex] / w;
+    const float coldZeroFraction = nodeCoverValues[coldIndex] / w;
+    float incomingZeroFraction   = 1.0f;
+    float incomingOneFraction    = 1.0f;
+
+    DAAL_ASSERT(hotZeroFraction < 1.0f);
+    DAAL_ASSERT(coldZeroFraction < 1.0f);
+
+    // see if we have already split on this feature,
+    // if so we undo that split so we can redo it for this node
+    std::uint32_t pathIndex = 0;
+    for (; pathIndex <= uniqueDepth; ++pathIndex)
+    {
+        if (uniquePath[pathIndex].featureIndex == splitFeature) break;
```

Comment:
You can check this thing right in the `for(;...;)` continuation condition

Reply:
It has become a bit more complex in some cases (added DAAL_ASSERTS). Would like to keep it consistent across different versions of TreeSHAP.